### PR TITLE
SPLAT-1218: AWS BYO VPC support for Wavelength Zones

### PR DESCRIPTION
**Pull request by mtulio**
This change introduces support for BYO VPC when installing a cluster in Wavelength Zones by:

- classifying subnets with zone type `wavelength-zone` as edge zone

This PR is created based on https://github.com/openshift/installer/pull/7369 , the following commits are related to this PR:

- [aws byovpc: classifying wavelength subnets as edge](https://github.com/openshift/installer/commit/cf9c3ac44d5286c781b3c9a1c1e3f77022b043d8)

- [aws/doc: updating edge zones Documentation with Wavelength Zone steps](https://github.com/openshift/installer/pull/7652/commits/77a07ffdd1d44bfaec3405d0df8ab75104488e65)

The following CloudFormation Templates have been introduced in https://github.com/openshift/installer/pull/7722:
- Create Carrier Gateway: provisions carrier gateway and route table for Wavelength Zones
- Create Public and Private subnets: create public and private subnets, associating each one to a provided route table

### CODE REVIEW

**Comment by mtulio** [comment]
Cluster successfully created when installing a cluter in existing VPC with subnets in Wavelength Zones:

- pre-req step: create VPC, carrier gateway and subnets

```bash
export CLUSTER_REGION=us-east-1
export CLUSTER_NAME=wlz-byon02

export CIDR_VPC="10.0.0.0/16"

export STACK_VPC=${CLUSTER_NAME}-vpc
aws cloudformation create-stack \
  --region ${CLUSTER_REGION} \
  --stack-name ${STACK_VPC} \
  --template-body file://$TEMPLATE_NAME_VPC \
  --parameters \
    ParameterKey=VpcCidr,ParameterValue="${CIDR_VPC}" \
    ParameterKey=AvailabilityZoneCount,ParameterValue=3 \
    ParameterKey=SubnetBits,ParameterValue=12

aws --region $CLUSTER_REGION cloudformation wait stack-create-complete --stack-name ${STACK_VPC}
aws --region $CLUSTER_REGION cloudformation describe-stacks --stack-name ${STACK_VPC}

export VPC_ID=$(aws --region $CLUSTER_REGION cloudformation describe-stacks \
  --stack-name ${STACK_VPC} \
  | jq -r '.Stacks[0].Outputs[] | select(.OutputKey=="VpcId").OutputValue' )

## Create WLZ Gateway
TEMPLATE_NAME_CARRIER_GW="./template-net-cagw.yaml"
ln -svf ~/go/src/github.com/mtulio/installer/upi/aws/cloudformation/01.01_carrier_gateway.yaml $TEMPLATE_NAME_CARRIER_GW

export STACK_CAGW=${CLUSTER_NAME}-cagw
aws cloudformation create-stack \
  --region ${CLUSTER_REGION} \
  --stack-name ${STACK_CAGW} \
  --template-body file://$TEMPLATE_NAME_CARRIER_GW \
  --parameters \
    ParameterKey=VpcId,ParameterValue="${VPC_ID}" \
    ParameterKey=ClusterName,ParameterValue="${CLUSTER_NAME}"

aws --region $CLUSTER_REGION cloudformation wait stack-create-complete --stack-name ${STACK_CAGW}
aws --region $CLUSTER_REGION cloudformation describe-stacks --stack-name ${STACK_CAGW}

# Setup prereq subnet
AZ_NAME="us-east-1-wl1-nyc-wlz-1"
AZ_SUFFIX=$(echo ${AZ_NAME/${CLUSTER_REGION}-/})

ZONE_GROUP_NAME=$(aws --region $CLUSTER_REGION ec2 describe-availability-zones \
  --filters Name=zone-name,Values=$AZ_NAME \
  | jq -r .AvailabilityZones[].GroupName)

aws --region $CLUSTER_REGION ec2 modify-availability-zone-group \
    --group-name "${ZONE_GROUP_NAME}" \
    --opt-in-status opted-in

export ROUTE_TABLE_PUB=$(aws --region $CLUSTER_REGION cloudformation describe-stacks \
  --stack-name ${STACK_CAGW} \
  | jq -r '.Stacks[0].Outputs[] | select(.OutputKey=="PublicRouteTableId").OutputValue' )


#> Select the first route table from the list
export ROUTE_TABLE_PVT=$(aws --region $CLUSTER_REGION cloudformation describe-stacks \
  --stack-name ${STACK_VPC} \
  | jq -r '.Stacks[0].Outputs[]
    | select(.OutputKey=="PrivateRouteTableIds").OutputValue
    | split(",")[0] | split("=")[1]' \
)

SUBNET_CIDR_PUB="10.0.128.0/24"
SUBNET_CIDR_PVT="10.0.129.0/24"

cat <<EOF
CLUSTER_REGION=$CLUSTER_REGION
VPC_ID=$VPC_ID
AZ_NAME=$AZ_NAME
AZ_SUFFIX=$AZ_SUFFIX
ZONE_GROUP_NAME=$ZONE_GROUP_NAME
ROUTE_TABLE_PUB=$ROUTE_TABLE_PUB
ROUTE_TABLE_PVT=$ROUTE_TABLE_PVT
SUBNET_CIDR_PUB=$SUBNET_CIDR_PUB
SUBNET_CIDR_PVT=$SUBNET_CIDR_PVT
EOF

# Subnet
TEMPLATE_NAME_SUBNET="./template-net-subnet.yaml"
ln -svf ~/go/src/github.com/mtulio/installer/upi/aws/cloudformation/01.99_subnet.yaml $TEMPLATE_NAME_SUBNET

export STACK_SUBNET=${CLUSTER_NAME}-subnets-${AZ_SUFFIX}
aws cloudformation create-stack \
  --region ${CLUSTER_REGION} \
  --stack-name ${STACK_SUBNET} \
  --template-body file://$TEMPLATE_NAME_SUBNET \
  --parameters \
    ParameterKey=VpcId,ParameterValue="${VPC_ID}" \
    ParameterKey=ClusterName,ParameterValue="${CLUSTER_NAME}" \
    ParameterKey=ZoneName,ParameterValue="${AZ_NAME}" \
    ParameterKey=PublicRouteTableId,ParameterValue="${ROUTE_TABLE_PUB}" \
    ParameterKey=PublicSubnetCidr,ParameterValue="${SUBNET_CIDR_PUB}" \
    ParameterKey=PrivateRouteTableId,ParameterValue="${ROUTE_TABLE_PVT}" \
    ParameterKey=PrivateSubnetCidr,ParameterValue="${SUBNET_CIDR_PVT}"

aws --region $CLUSTER_REGION cloudformation wait stack-create-complete --stack-name ${STACK_SUBNET}
aws --region $CLUSTER_REGION cloudformation describe-stacks --stack-name ${STACK_SUBNET}


## Get the subnets

# Public Subnets from VPC Stack
mapfile -t SUBNETS < <(aws --region $CLUSTER_REGION cloudformation describe-stacks   --stack-name "${STACK_VPC}" --query "Stacks[0].Outputs[?OutputKey=='PrivateSubnetIds'].OutputValue" --output text | tr ',' '\n')

# Private Subnets from VPC Stack
mapfile -t -O "${#SUBNETS[@]}" SUBNETS < <(aws --region $CLUSTER_REGION cloudformation describe-stacks   --stack-name "${STACK_VPC}" --query "Stacks[0].Outputs[?OutputKey=='PublicSubnetIds'].OutputValue" --output text | tr ',' '\n')

echo ${SUBNETS[@]}

# SELECT ONE subnet

# Public Edge Subnet
#mapfile -t -O "${#SUBNETS[@]}" SUBNETS < <(aws --region $CLUSTER_REGION cloudformation describe-stacks   --stack-name "${STACK_SUBNET}" --query "Stacks[0].Outputs[?OutputKey=='PublicSubnetIds'].OutputValue" --output text | tr ',' '\n')

# Private Edge Subnet
mapfile -t -O "${#SUBNETS[@]}" SUBNETS < <(aws --region $CLUSTER_REGION cloudformation describe-stacks   --stack-name "${STACK_SUBNET}" --query "Stacks[0].Outputs[?OutputKey=='PrivateSubnetIds'].OutputValue" --output text | tr ',' '\n')

echo ${SUBNETS[@]}


## Cluster config

export PULL_SECRET_FILE=${HOME}/.openshift/pull-secret-latest.json
export BASE_DOMAIN=devcluster.openshift.com
export SSH_PUB_KEY_FILE=$HOME/.ssh/id_rsa.pub

CLUSTER_NAME_VARIANT=${CLUSTER_NAME}-0
INSTALL_DIR=${CLUSTER_NAME_VARIANT}
mkdir $INSTALL_DIR

cat <<EOF > ${INSTALL_DIR}/install-config.yaml
apiVersion: v1
publish: External
baseDomain: ${BASE_DOMAIN}
metadata:
  name: "${CLUSTER_NAME_VARIANT}"
platform:
  aws:
    region: ${CLUSTER_REGION}
    subnets:
$(for SB in ${SUBNETS[*]}; do echo "    - $SB"; done)
pullSecret: '$(cat ${PULL_SECRET_FILE} | awk -v ORS= -v OFS= '{$1=$1}1')'
sshKey: |
  $(cat ${SSH_PUB_KEY_FILE})
EOF

cp ${INSTALL_DIR}/install-config.yaml ${INSTALL_DIR}/install-config.yaml.bkp


./openshift-install create manifests --dir $INSTALL_DIR

./openshift-install create cluster --dir $INSTALL_DIR --log-level=debug
```

- install-config
```sh
$ oc get cm -n kube-system cluster-config-v1 -o json | jq -r '.data["install-config"]' | yq .platform
{
  "aws": {
    "region": "us-east-1",
    "subnets": [
      "subnet-0687b3d518c58e6b7",
      "subnet-037220f93711e24b8",
      "subnet-0a2392f020093519d",
      "subnet-04323e9fad756c3f2",
      "subnet-0422ef1adaf65bcea",
      "subnet-0de2f77bd4b7cab64",
      "subnet-000af0974290dce77"
    ]
  }
}

```

- cluster-info

```
$ oc get clusterversion
NAME      VERSION       AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.15.0-ec.1   True        False         20m     Cluster version is 4.15.0-ec.1

$ oc get machines -n openshift-machine-api
NAME                                                    PHASE     TYPE         REGION      ZONE                      AGE
wlz-byon02-0-phk6l-edge-us-east-1-wl1-nyc-wlz-1-ft26w   Running   r5.2xlarge   us-east-1   us-east-1-wl1-nyc-wlz-1   29m
wlz-byon02-0-phk6l-master-0                             Running   m6i.xlarge   us-east-1   us-east-1a                34m
wlz-byon02-0-phk6l-master-1                             Running   m6i.xlarge   us-east-1   us-east-1b                34m
wlz-byon02-0-phk6l-master-2                             Running   m6i.xlarge   us-east-1   us-east-1c                34m
wlz-byon02-0-phk6l-worker-us-east-1a-2srj7              Running   m6i.xlarge   us-east-1   us-east-1a                29m
wlz-byon02-0-phk6l-worker-us-east-1b-hx9dn              Running   m6i.xlarge   us-east-1   us-east-1b                29m
wlz-byon02-0-phk6l-worker-us-east-1c-2dkgz              Running   m6i.xlarge   us-east-1   us-east-1c                29m

$ oc get nodes
NAME                          STATUS   ROLES                  AGE   VERSION
ip-10-0-129-32.ec2.internal   Ready    edge,worker            29m   v1.28.2+481304a
ip-10-0-48-41.ec2.internal    Ready    control-plane,master   40m   v1.28.2+481304a
ip-10-0-63-85.ec2.internal    Ready    worker                 33m   v1.28.2+481304a
ip-10-0-67-95.ec2.internal    Ready    worker                 33m   v1.28.2+481304a
ip-10-0-76-141.ec2.internal   Ready    control-plane,master   40m   v1.28.2+481304a
ip-10-0-81-211.ec2.internal   Ready    worker                 33m   v1.28.2+481304a
ip-10-0-86-205.ec2.internal   Ready    control-plane,master   40m   v1.28.2+481304a

```

**Comment by mtulio** [comment]
PR is ready for review. @patrickdillon can I have the installer help review this PR in the next sprint?
/assign @patrickdillon

**Comment by mtulio** [comment]
PR rebased. Converting to draft to improve the commit descriptions.

**Comment by mtulio** [comment]
/pj-rehearse pull-ci-openshift-installer-master-e2e-aws-ovn-shared-vpc-wavelengthzones

**Comment by mtulio** [comment]
[Presubmit merged](https://github.com/openshift/release/pull/45158), and green for `e2e-aws-ovn-shared-vpc-wavelengthzones`. PR ready for review. Could you please take a look?
/assign @r4f4 @patrickdillon 
cc @makentenza

**Comment by mtulio** [comment]
@yunjiang29 - with regards the [SPLAT-1255](https://issues.redhat.com/browse/SPLAT-1255), the following presubmit is testing the feature: `e2e-aws-ovn-shared-vpc-wavelengthzones`.

**Comment by mtulio** [comment]
@patrickdillon @r4f4 , after discussing some use cases with @yunjiang29 , I have confirmed one very specific scenario which was not covered in this PR:
- Installing a cluster in existing VPC with Wavelength subnets associated to a public "edge" route table (has default route the Carrier Gateway)

This scenario will not work without https://github.com/openshift/machine-api-provider-aws/pull/78 , for that reason we need to exercise creating a custom release including that PR - this is another reason we are not exercising yet in CI for those WL PRs. There we are testing using patch manifest strategy, which was bypassing this issue.

As this PR wasn't merged, I provided the fix [here](https://github.com/openshift/installer/pull/7652/files#diff-1f9c4b4d491457121c02e90efed5c7de4167938a95e2cfd11c7273327829b289R232-R234) .

File: pkg/asset/installconfig/aws/subnet.go [hunk 1]
```diff
@@ -164,7 +164,8 @@ func subnets(ctx context.Context, session *session.Session, region string, ids [
 		}
 
 		// AWS Local Zones are grouped as Edge subnets
-		if meta.Zone.Type == typesaws.LocalZoneType {
+		if meta.Zone.Type == typesaws.LocalZoneType ||
+			meta.Zone.Type == typesaws.WavelengtyZoneType {
```
Comment (patrickdillon) [inline pkg/asset/installconfig/aws/subnet.go:168]: ooh, I just noticed it looks like we merged (in a previous commit) a typo here?

```suggestion
			meta.Zone.Type == typesaws.WavelengthZoneType {
```
we could fix here or in another pr, either is fine

**Review by patrickdillon — COMMENTED** [review]
/approve

**Pull request merged**